test/e2e/storage/host_path_type.go

```diff
@@ -0,0 +1,332 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"fmt"
+	"path"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = utils.SIGDescribe("HostPathType", func() {
+	f := framework.NewDefaultFramework("host-path-type")
+
+	var (
+		ns           string
+		hostBaseDir  string
+		mountBaseDir string
+		basePod      *v1.Pod
+
+		hostPathUnset             = v1.HostPathUnset
+		hostPathDirectoryOrCreate = v1.HostPathDirectoryOrCreate
+		hostPathDirectory         = v1.HostPathDirectory
+		hostPathFileOrCreate      = v1.HostPathFileOrCreate
+		hostPathFile              = v1.HostPathFile
+		hostPathSocket            = v1.HostPathSocket
+		hostPathCharDev           = v1.HostPathCharDev
+		hostPathBlockDev          = v1.HostPathBlockDev
+	)
+
+	BeforeEach(func() {
+		ns = f.Namespace.Name
+
+		By("Create a pod for further testing")
+		hostBaseDir := path.Join("/tmp", ns)
+		mountBaseDir := "/mnt/test"
+		basePod = f.PodClient().CreateSync(newHostPathTypeTestPod("", hostBaseDir, mountBaseDir, &hostPathDirectoryOrCreate))
+	})
+
+	Context("HostPathType Directory Test", func() {
+		var targetDir string
+
+		BeforeEach(func() {
+			targetDir = path.Join(hostBaseDir, "adir")
+			By("Should automatically create a new directory 'adir' when HostPathType is HostPathDirectoryOrCreate")
+			newPod, err := f.ClientSet.CoreV1().Pods(ns).Create(newHostPathTypeTestPod(basePod.Spec.NodeName, targetDir, "/mnt/test", &hostPathDirectoryOrCreate))
+			Expect(err).NotTo(HaveOccurred())
+			framework.ExpectNoError(framework.WaitForPodNameRunningInNamespace(f.ClientSet, newPod.Name, newPod.Namespace))
+		})
+
+		It("Should fail on mounting non-existent directory 'does-not-exist-dir' when HostPathType is HostPathDirectory", func() {
+			dirPath := path.Join(hostBaseDir, "does-not-exist-dir")
+			testPodHostPathType(f, basePod.Spec.NodeName, dirPath, fmt.Sprintf("%s is not a directory", dirPath), &hostPathDirectory)
```

Comment:
It looks like `testPodHostPathType` is used for failure scenarios. Maybe a better name could be `verifyPodHostPathTypeFailure`